chainer/sequential.py

```diff
@@ -0,0 +1,405 @@
+                'given {} is not callable.'.format(layer))
+
+        self._layers.insert(i, layer)
+        if isinstance(layer, chainer.Link):
+            self.add_link(layer)
+
+    def remove(self, layer):
+        if layer in self:
+            del self[self.index(layer)]
+        else:
+            raise ValueError(
+                'There is no layer object that is same as {}'.format(layer))
+
+    def remove_by_layer_type(self, type_name):
+        """Remove layers by layer type.
+
+        This method removes layers from the Sequential object by the
+        layer's class name or function name. If you want to remove a
+        :class:`~Link`, the argment ``type_name`` should be its class name,
+        e.g., :class:`~links.Linear` or :class:`~links.Convolution2D`, etc.
+        If you want to remove a :class:`~Function` class or any other callable
+        objects, ``type_name`` should be the function name, e.g., ``relu`` or
+        ``reshape``, etc.
+
+        Args:
+            type_name (str): The name of a layer you want to remove.
+
+        """
+
+        names = []
+        for layer in self:
+            if isinstance(layer, chainer.Link):
+                name = layer.__class__.__name__
+            else:
+                name = layer.__name__
+            names.append((name, layer))
+        for _name, _layer in names:
+            if type_name == _name:
+                self.remove(_layer)
+
+    def pop(self, i=-1):
+        layer = self._layers[i]
+        del self[i]
+        return layer
+
+    def clear(self):
+        for i, link in enumerate(self._children):
+            del self._children[i]
+        self._layers = []
+
+    def index(self, layer, start=None, end=None):
+        return self._layers[start:end].index(layer)
+
+    def count(self, layer):
+        return self._layers.count(layer)
+
+    def count_by_layer_type(self, type_name):
+        """Count the number of layers by layer type.
+
+        This method counts the number of layers which have the name given by
+        the argment ``type_name``. For example, if you want to know the number
+        of :class:`~links.Linear` layers included in this model, ``type_name``
+        should be ``Linear``. If you want to know the number of
+        :class:`~Function` classes or user-defined functions which have a
+        specific name, ``type_name`` should be the function name, e.g.,
+        ``relu`` or ``reshape``, etc.
+
+        Args:
+            type_name (str): The class or function name of a layer you want to
+                enumerate.
+
+        """
+
+        num = 0
+        for layer in self._layers:
+            if isinstance(layer, chainer.Link):
+                if layer.__class__.__name__ == type_name:
+                    num += 1
+            else:
+                if layer.__name__ == type_name:
+                    num += 1
+        return num
+
+    def sort(self, key=None, reverse=None):
+        raise NotImplementedError
```

Comment:
`NotImplementedError` in Python means "this method should be overridden by a subclass", so it's not appropriate to use it here. Consider another exception type, or just remove this definition (ditto for `reverse`).

Reply:
OK, I remove both.

Reply:
Hmm, but, for example, `Variable`'s `__eq__` raises `NotImplementedError`, so it's intended to be overridden by a subclass?

Reply:
Thanks!